### ISSUE TITLE
Is there a way to obtain the HTML currently being rendered in the WebView?

### ISSUE THREAD

**Issue by uglyluigi** [Boscop/web-view#182]
This would be useful for use in a 3rd party website, where you need to use a WebView to render the web app and then capture the HTML for processing. Is there a way to do this?

**Comment by edko99** [comment]
First, you can use JS to get the whole document as a string (here is how: https://stackoverflow.com/questions/817218/how-to-get-the-entire-document-html-as-a-string).

After that, if you want, you can pass that string to the web-view library, via `invoke` and process it in Rust.

**Comment by uglyluigi** [comment]
Thanks a lot!

**Comment by uglyluigi** [comment]
I'm in a place where I can properly capture the HTML, but when using the Edge engine for web rendering at least, both document.documentElement.inner/outerHTML appear to be undefined or null.

The XMLSerializer solution listed in your SO article doesn't work, but at least returns an empty string.

`extern crate web_view;

use web_view::*;

pub fn init() -> Result<(), reqwest::Error> {
	println!("Connecting to turnip exchange...");
	
	let mut view = WebViewBuilder::new()
        .title("Minimal webview example")
        .content(Content::Url("https://turnip.exchange/"))
        .size(800, 600)
        .resizable(true)
        .debug(true)
        .user_data(())
        .invoke_handler(|_webview, _arg| {
			println!("Arg is: {}" , _arg);
			
			if let _arg = "getHtml" {
				_webview.eval("external.invoke('()');");
			}
			
			Ok(())
		})
        .build()
        .unwrap();

	match view.eval("external.invoke('getHtml');") {
		Ok(()) => println!("Eval went ok"),
		Err(err) => println!("Eval error: {}" , err),
	};
	
	view.run();

	//let mut content = view.content;
	
	Ok(())
}`

I suspect it may be because I'm trying to obtain the HTML before I call `run()` on the view, but I haven't devised a way to use `eval()` on the view after calling run() since after calling `run()` the view is moved.

What do you suggest?

**Comment by edko99** [comment]
I think you are right: you are calling `eval` before you call `run`.

I imagine that your issue is that you don't have control over the webpage's content, and it is created dynamically with JS, so an http GET from its source URL won't do.

Looking at the WebView source code, I see there is a pub function `handle`, which "creates a thread-safe Handle to the WebView, from which closures can be dispatched". I have no experience with it, but you may want to look into it. It might let you call into JS from another thread, some time after you call `run`.

**Comment by uglyluigi** [comment]
Yeah I realized I had quotes after I edited my comment, but rest assured that those were not present when I was testing the suggestions in the SO thread. I'll have to give that a shot, thanks!

**Comment by uglyluigi** [comment]
This is what I came up with:

```
extern crate web_view;

use web_view::*;
use std::thread;
use std::time::Duration;

pub fn init() -> Result<(), reqwest::Error> {
	println!("Connecting to turnip exchange...");
	
	let mut view = WebViewBuilder::new()
        .title("Minimal webview example")
        .content(Content::Url("https://turnip.exchange/"))
        .size(800, 600)
        .resizable(true)
        .debug(true)
        .user_data(())
        .invoke_handler(|_webview, _arg| {
			println!("Arg is: {}" , _arg);
			
			if let _arg = "getHtml" {
				_webview.eval("external.invoke(document.documentElement.outerHTML);");
			}
			
			Ok(())
		})
        .build()
        .unwrap();

	let view_handle = view.handle();
	view.run();
	
	let handle = thread::spawn(move || {
		println!("Running sleeper thread");
		view_handle.dispatch(|view| {
			match view.eval("external.invoke('getHtml');") {
			Ok(()) => {
				println!("Eval went ok");
				Ok(())
			},
			Err(err) => {
				println!("Eval error: {}" , err);
				Err(err)
			},
		}
		});
	});
	
	
	handle.join();
	
	
	Ok(())
}
```

I have to say, `run()` moving the web view out of scope is incredibly inconvenient for what I'm trying to accomplish. I'm not sure what I'm trying to do is even possible, as `dispatch()` simply schedules a closure for execution but there's no way to step the event thread after calling `run()` so it's never executed. I managed to capture the HTML that's served from the GET request, but if that's what I needed to do I would've just used reqwest. Not sure where to go from here! Any suggestions?

**Comment by edko99** [comment]
Maybe you could spawn the thread before calling `run`? Within the thread, you could introduce a small delay before dispatching.

**Comment by uglyluigi** [comment]
So, to my understanding, `dispatch` simply queues a callback for execution when you call `step` or `run`. Unfortunately, if a callback is queued with `dispatch` and then `run` is called, there's no way to run additional callbacks at that point. At least with how I understand the lifecycle of a WebView object.

**Comment by edko99** [comment]
I don't think that is the case. Please check the `timer.rs` program in the examples directory.
